ethcore/src/tx_filter.rs

```diff
@@ -60,36 +63,48 @@ impl TransactionFilter {
 	/// Check if transaction is allowed at given block.
 	pub fn transaction_allowed<C: BlockInfo + CallContract>(&self, parent_hash: &H256, transaction: &SignedTransaction, client: &C) -> bool {
 		let mut cache = self.permission_cache.lock();
-
-		let tx_type = match transaction.action {
-			Action::Create => tx_permissions::CREATE,
+		let (tx_type, to) = match transaction.action {
+			Action::Create => (tx_permissions::CREATE, Address::new()),
 			Action::Call(address) => if client.code_hash(&address, BlockId::Hash(*parent_hash)).map_or(false, |c| c != KECCAK_EMPTY) {
-				tx_permissions::CALL
-			} else {
-				tx_permissions::BASIC
-			}
+					(tx_permissions::CALL, address)
+				} else {
+					(tx_permissions::BASIC, address)
+				}
 		};
 
 		let sender = transaction.sender();
-		let key = (*parent_hash, sender);
+		let value = transaction.value;
+		let key = (*parent_hash, sender, to, value);
```

Comment:
Just realized that it makes the cache pretty useless and the whole transaction filter checking way slower - every transaction will cause a state call anyway (cause it's quite unlikely all 3 parameters will be the same).

Perhaps we should consider either:
1. removing the cache completely
2. change the contract return type and allow some quantification (like: permission `None`, sender never allowed)